Provide a decision tree for artifacts

sudo-bmitch opened the pull request:
This adds a decision tree to the artifact guidance to indicate when the artifactType field should and should not be used. I believe this closes #1100.

tianon commented on artifacts-guidance.md:
```diff
@@ -1,5 +1,5 @@
 Content other than OCI container images MAY be packaged using the image manifest.
-When this is done, the `config.mediaType` value MUST be set to a value specific to the artifact type or the [empty value](manifest.md#guidance-for-an-empty-descriptor).
-Additional details and examples are provided in the [image manifest specification](manifest.md#guidelines-for-artifact-usage).
+When this is done, the `config.mediaType` value will not be a known OCI image config [media type](media-types.md).
```
Hmm, isn't it totally valid for a non-image artifact to set `artifactType` but still use an OCI media type for config? (for example, a fully compliant image config)

I don't see an obvious use case for doing so, but this feels like an unnecessary restriction to me.

sudo-bmitch replied:
In that scenario, how would a container runtime differentiate between a container image and something the runtime should ignore?

sudo-bmitch replied:
I believe what buildkit is doing would politely be referred to as a hack or kludge to bypass registry filters. There's a [PR to fix it](https://github.com/moby/buildkit/pull/3610) but it's blocked because at least one registry refuses unknown/non-image content. The [config spec](https://github.com/opencontainers/image-spec/blob/main/config.md) describes it as:

> This specification outlines the JSON format describing images for use with a container runtime and execution tool and its relationship to filesystem changesets, described in [Layers](https://github.com/opencontainers/image-spec/blob/main/layer.md).

I hesitate to recommend other uses that may cause issues for existing runtimes that don't know about the `artifactType` field.

sudo-bmitch replied:
From today's call, we added a note to clarify that some existing artifacts did not follow this OCI advice, and softened the "will" to a "should" even though this part is guidance and not treated the same as the other parts of the spec.

tianon replied:
Just to clarify, I wasn't using BuildKit's objects as an example of "a good artifact", but rather as an example object that does currently exist today, and which runtimes already have to deal with in some way (and I think already probably deal with in the only way they reasonably can, by balking at the layer media types).

(I also suggested `artifactType` to them in the PR you linked :smile:)

tianon commented on manifest.md:
```diff
@@ -175,6 +175,14 @@ If the `config.mediaType` is set to the empty value, the `artifactType` MUST be
+
+1. Does the artifact consist of at least one file or blob? If yes, continue to 2. If no, specify the `artifactType`, and set the `config` and `layers` to their empty descriptor values.
+2. Does the artifact have additional JSON formatted metadata as configuration? If yes, continue to 3. If no, specify the `artifactType`, include the artifact in the `layers`, and set `config` to the empty descriptor value.
+3. Does the artifact require a different `artifactType` from the `config.mediaType`, to support multiple configuration formats or versions for the same artifact tooling? If yes, continue to 4. If no, specify the `config` with the metadata, include the artifact in the `layers`, but do not set the `artifactType`.
```
This one has me confused; isn't the preference for artifacts to _always_ set `artifactType` except for cases of backwards compatibility (ie, where the format existed before `artifactType` was defined)?

sudo-bmitch replied:
I'm not opposed to that, but it feels redundant.

tianon replied:
It is a little redundant, but given this is recommendations specifically for new artifact authors, don't we want to end with a world where _most_ non-image artifacts are setting `artifactType` consistently and the config media type fallback is rarely invoked? (Thinking especially about registry implementors here, where a consistent data model is more helpful than being slightly less redundant.)

sudo-bmitch replied:
In the cases where existing artifacts are encountered (e.g. helm charts before we convince them to change their tooling), are we suggesting that registry implementers will eventually drop their support for those artifacts? Can they drop that support as soon as 1.1 is released?

sudo-bmitch replied:
From today's call we consolidated the last two options, always recommending `artifactType`, but with a note at the end to clarify that existing artifacts do not do this.

tianon reviewed (changes requested):
(explicit NACK, for the avoidance of doubt)

ChaoyiHuang commented on manifest.md:
```diff
@@ -175,6 +175,14 @@ If the `config.mediaType` is set to the empty value, the `artifactType` MUST be
+The decision tree below MAY be used to structure your artifact:
+
+1. Does the artifact consist of at least one file or blob? If yes, continue to 2. If no, specify the `artifactType`, and set the `config` and a single `layers` element to the empty descriptor value.
+2. Does the artifact have additional JSON formatted metadata as configuration? If yes, continue to 3. If no, specify the `artifactType`, include the artifact in the `layers`, and set `config` to the empty descriptor value.
```
For non-image artifact, does it allow this usage, the layer descriptor in the layers pointing to another manifest, i.e, the media type in the layer is "application/vnd.oci.image.manifest.v1+json".

In current image-spec/manifest.md description, it does not set the restriction on this usage.

sudo-bmitch replied:
I believe that's a separate issue/question from this PR, so it would be better to raise that on its own for visibility.

My own take is that the spec doesn't say the media type allowed for the layer, only what container runtimes should support. However, I think you'll encounter issues pushing that to a registry and handled by generic tools because the layers field content is treated as an opaque blob, not necessarily processed to extract content like a manifest. Blobs are pushed to a separate API on the registry, registries may reject a manifest if all blobs are not pushed, and a manifest pushed as a blob may not be processed by the registry to retain embedded content from the garbage collector.

If you want a manifest and references other manifests, then you likely want an OCI Index.

sudo-bmitch commented:
This has been updated with feedback from today's call. I've also incorporated @dasiths's suggestions from #1100 to inline the examples and added credit in the commit message.

shizhMSFT reviewed (approved):
LGTM but IANAM

sajayantony commented:
@tianon PTAL. 
I think it might be good for this guidance to be released as a part of v1.1 
Planning to send out a vote as per #1093

tianon commented on artifacts-guidance.md:
```diff
@@ -1,5 +1,6 @@
 Content other than OCI container images MAY be packaged using the image manifest.
-When this is done, the `config.mediaType` value MUST be set to a value specific to the artifact type or the [empty value](manifest.md#guidance-for-an-empty-descriptor).
-Additional details and examples are provided in the [image manifest specification](manifest.md#guidelines-for-artifact-usage).
+When this is done, the `config.mediaType` value should not be a known OCI image config [media type](media-types.md).
+Historically, due to registry limitations, some tools have created non-OCI conformant artifacts using the `application/vnd.oci.image.config.v1+json` value for `config.mediaType` and values specific to the artifact in `layer[*].mediaType`.
```
:+1: on the updated language here :heart:
